### PR TITLE
Implement MSPileup clean-up task

### LINKED ISSUE #11514: [MS] Redefine MSPileup document deletion logic
**Impact of the new feature**
MSPileup

**Is your feature request related to a problem? Please describe.**
While discussing about this service with Hasan and Todor today, we noticed that we cannot really trigger a document deletion upon a DELETE request. 

Reason for that is very simple - and overlooked - if a document is removed from the database backend, then there is nothing that the service can act on for that pileup. In other words, we would be potentially leaving Rucio rules behind and we would never be able to clean those up automatically. Hence, we need to redefine that logic, see below.
 
**Describe the solution you'd like**
Upon a DELETE HTTP request, MSPileup should not really delete the relevant pileup object. Instead, it should update the following pileup attributes:
* `deactivatedOn`: the server will define it according to the service documentation (thus an integer with the epoch time)
* `active`: False (boolean)

which will then allow MSPileup to fetch it in the next service polling cycle and properly delete any outstanding Rucio rule.
 
**Describe alternatives you've considered**
Remove the HTTP DELETE method, thus no longer support it, and enforce users to update the document accordingly (when a deletion is required).

**Additional context**
Related to: https://github.com/dmwm/WMCore/issues/11425

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes #11514 

#### Status
ready

#### Description
Implement MSPileup clean-up task

#### Is it backward compatible (if not, which system it affects?)
YES

### CODE REVIEW

**Comment by vkuznet** [comment]
Alan, this PR is small and ready for review. There are couple of unstable test failure which are not related to code in question though.

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py [hunk 1]
```diff
@@ -62,6 +62,27 @@ def pileupSizeTask(self):
...
+           document has been deactivated for a while (deactivatedOn=XXX),
+        2. For those documents which are fetched make delete call to backend database
+        """
+        spec = {'active': False}
+        docs = self.mgr.getPileup(spec)
+        deleteDocs = 0
+        for doc in docs:
+            if not doc['ruleIds'] and not doc['currentRSEs']:
+                spec = {'_id': doc['_id']}
+                self.mgr.deletePileup(spec)
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:82]: Can you please make a new log record whenever a pileup object is deleted (saying which pileup object was deleted)?

**Review by amaltaro — CHANGES_REQUESTED** [review]
@vkuznet Valentin, other than one comment inline, the code looks good and simple. However, I think we do need to have the following functionality here:
1) have a service configurable parameter specifying the number of days since the pileup object has been deactivated;
2) use this configurable parameter and account for `deactivatedOn` in the pileup deletion as well, something like `if (time - deactivatedOn) > X days: then delete`

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py (hunk 1 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:86]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py [hunk 2]
```diff
@@ -62,6 +62,31 @@ def pileupSizeTask(self):
             msg = f"MSPileup pileup size task failed with error {exp}"
             self.logger.exception(msg)
 
+    def cleanupTask(self, timeThreshold=15):
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:65]: I think this argument (timeThreshold) should be mandatory.

File: src/python/WMCore/MicroService/MSPileup/MSPileupTaskManager.py [hunk 3]
```diff
@@ -33,6 +33,7 @@ def __init__(self, msConfig, **kwargs):
         self.rucioAccount = msConfig.get('rucioAccount', 'wmcore_transferor')
         self.rucioUrl = msConfig['rucioUrl']  # aligned with MSCore init
         self.rucioAuthUrl = msConfig['rucioAuthUrl']  # aligned with MSCore init
+        self.cleanupTimeThreshold - msConfig.get('cleanupTimeThreshold', 15)
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupTaskManager.py:36]: typo. Please also add a comment saying that it's expected to be in days. Or maybe we should call it `cleanupDaysThreshold`? Up to you.

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py [hunk 4]
```diff
@@ -62,6 +62,31 @@ def pileupSizeTask(self):
...
+           - empty currentRSEs; and
+           document has been deactivated for a while (deactivatedOn=XXX),
+        2. For those documents which are fetched make delete call to backend database
+
+        :param timeThreshold: time threshold in days which will determine document clean-up readiness
+        """
+        spec = {'active': False}
+        docs = self.mgr.getPileup(spec)
+        deleteDocs = 0
+        thr = timeThreshold * 24 * 60 * 60  # convert to second
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:81]: I know you can give `thr` a more meaningful name ;)

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, jenkins report points to a clear error in the code. Please do always take advantage of the automated report.

In addition to that, I left a few comments along the code and would like to ask you to also update the production service configuration to delete pileup after 60 days (I think it's a good commitment).

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py (hunk 2 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:65]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupTaskManager.py (hunk 3 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupTaskManager.py:36]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py (hunk 4 above)
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:81]: done

**Review by amaltaro — APPROVED** [review]
It looks good to me. However, can you please:
1) provide the relevant service configuration parameter in this MR https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/194#note_6550577
2) update the MSPileup wiki with this additional task description

I have a remark to make though, related to the polling cycle of this clean up logic. If we wanted to save CPU cycles, we could come up with a logic that would execute it only once or twice a day. But I like the simplicity that it gives by just making it part of the usual cycle and I do not think it's an expensive task. So we should be all good!

**Comment by vkuznet** [comment]
There is no need to worry about extra CPU cycles since over here and clean-up task should go along with other tasks. I'll update wiki and provide config parameters but please merge this PR as it blocks another one https://github.com/dmwm/WMCore/pull/11478.

**Comment by vkuznet** [comment]
wiki has been updated, i.e. I added clean-up task description.

**Comment by amaltaro** [comment]
FYI Production configuration is provided in this PR: https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/202

**Pull request merged**